ChannelAtomicCancelStressTest fails with CoroutinesInternalError caused by ClassCastException

https://teamcity.jetbrains.com/viewLog.html?buildId=2565122&buildTypeId=KotlinTools_KotlinxCoroutines_BuildWindows
`ChannelAtomicCancelStressTest.testAtomicCancelStress[RendezvousChannel]`
```
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[ThreadPoolDispatcher[2, ChannelAtomicCancelStressTest], Continuation at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchReceiver$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:135)@7015c908]){CancelledContinuation[kotlinx.coroutines.JobCancellationException: Job was cancelled; job="coroutine#563998":StandaloneCoroutine{Cancelled}@7ed2297a]}@7aeda278. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:86)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.DispatchedContinuation
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:106)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:192)
	... 7 more
```

https://teamcity.jetbrains.com/viewLog.html?buildId=2555599&buildTypeId=KotlinTools_KotlinxCoroutines_BuildWindows
`ChannelAtomicCancelStressTest.testAtomicCancelStress[RendezvousChannel]`
```
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[ThreadPoolDispatcher[2, ChannelAtomicCancelStressTest], Continuation at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchSender$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:108)@54cd1478]){CancelledContinuation[kotlinx.coroutines.JobCancellationException: Job was cancelled; job="coroutine#567226":StandaloneCoroutine{Cancelled}@11cf9085]}@6477608b. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:86)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.DispatchedContinuation
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:106)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:192)
	... 7 more
```

https://teamcity.jetbrains.com/viewLog.html?buildId=2556063&tab=buildResultsDiv&buildTypeId=KotlinTools_KotlinxCoroutines_Build#testNameId-3299188611556527194
`ChannelAtomicCancelStressTest.testAtomicCancelStress[RendezvousChannel]`
```
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[ThreadPoolDispatcher[2, ChannelAtomicCancelStressTest], Continuation at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchSender$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:108)@3ea5b47b]){CancelledContinuation[kotlinx.coroutines.JobCancellationException: Job was cancelled; job="coroutine#897075":StandaloneCoroutine{Cancelled}@7b38ff4b]}@50e96830. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:92)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:63)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.DispatchedContinuation
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:106)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:187)
	... 7 more
```

https://teamcity.jetbrains.com/viewLog.html?buildId=2554902&tab=buildResultsDiv&buildTypeId=KotlinTools_KotlinxCoroutines_BuildWindows#testNameId-3299188611556527194
`ChannelAtomicCancelStressTest.testAtomicCancelStress[RendezvousChannel]`
```
ava.lang.AssertionError
	at kotlinx.coroutines.channels.SendElement.tryResumeSend(AbstractChannel.kt:1055)
	at kotlinx.coroutines.channels.AbstractChannel$TryPollDesc.onPrepare(AbstractChannel.kt:684)
	at kotlinx.coroutines.internal.LockFreeLinkedListNode$PrepareOp.perform(LockFreeLinkedList.kt:404)
	at kotlinx.coroutines.internal.LockFreeLinkedListNode$AbstractAtomicDesc.prepare(LockFreeLinkedList.kt:472)
	at kotlinx.coroutines.selects.SelectBuilderImpl$AtomicSelectOp.prepare(Select.kt:567)
	at kotlinx.coroutines.internal.AtomicOp.perform(Atomic.kt:89)
	at kotlinx.coroutines.selects.SelectBuilderImpl.performAtomicTrySelect(Select.kt:541)
	at kotlinx.coroutines.channels.AbstractChannel.pollSelectInternal(AbstractChannel.kt:527)
	at kotlinx.coroutines.channels.AbstractChannel.registerSelectReceive(AbstractChannel.kt:705)
	at kotlinx.coroutines.channels.AbstractChannel.access$registerSelectReceive(AbstractChannel.kt:485)
	at kotlinx.coroutines.channels.AbstractChannel$onReceive$1.registerSelectClause1(AbstractChannel.kt:694)
	at kotlinx.coroutines.selects.SelectBuilderImpl.invoke(Select.kt:616)
	at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchReceiver$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:136)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.selects.SelectBuilderImpl.resumeWith(Select.kt:280)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:49)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:439)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
kotlinx.coroutines.CompletionHandlerException: Exception in cancellation handler for CancellableContinuation(DispatchedContinuation[ThreadPoolDispatcher[2, ChannelAtomicCancelStressTest], Continuation at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchSender$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:108)@72d8978c]){CancelledContinuation[kotlinx.coroutines.JobCancellationException: Job was cancelled; job="coroutine#564106":StandaloneCoroutine{Cancelling}@69c5a33b]}@2e5facbd
	at kotlinx.coroutines.CancellableContinuationImpl.cancel(CancellableContinuationImpl.kt:478)
	at kotlinx.coroutines.CancellableContinuationImpl.parentCancelled$kotlinx_coroutines_core(CancellableContinuationImpl.kt:181)
	at kotlinx.coroutines.ChildContinuation.invoke(JobSupport.kt:1429)
	at kotlinx.coroutines.ChildContinuation.invoke(JobSupport.kt:1424)
	at kotlinx.coroutines.JobSupport.invokeOnCompletion(JobSupport.kt:1493)
	at kotlinx.coroutines.Job$DefaultImpls.invokeOnCompletion$default(Job.kt:325)
	at kotlinx.coroutines.CancellableContinuationImpl.setupCancellation(CancellableContinuationImpl.kt:115)
	at kotlinx.coroutines.CancellableContinuationImpl.getResult(CancellableContinuationImpl.kt:226)
	at kotlinx.coroutines.channels.AbstractSendChannel.sendSuspend(AbstractChannel.kt:1108)
	at kotlinx.coroutines.channels.AbstractSendChannel.send(AbstractChannel.kt:149)
	at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchSender$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:108)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.selects.SelectBuilderImpl.resumeWith(Select.kt:280)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:49)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:439)
	at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:303)
	at java.util.concurrent.FutureTask.run(FutureTask.java:138)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:98)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:206)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:895)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:918)
	at java.lang.Thread.run(Thread.java:662)
Caused by: java.lang.AssertionError
	at kotlinx.coroutines.channels.SendElement.tryResumeSend(AbstractChannel.kt:1055)
	at kotlinx.coroutines.channels.AbstractChannel$TryPollDesc.onPrepare(AbstractChannel.kt:684)
	at kotlinx.coroutines.internal.LockFreeLinkedListNode$PrepareOp.perform(LockFreeLinkedList.kt:404)
	at kotlinx.coroutines.internal.LockFreeLinkedListNode.getNext(LockFreeLinkedList.kt:92)
	at kotlinx.coroutines.internal.LockFreeLinkedListNode.remove(LockFreeLinkedList.kt:238)
	at kotlinx.coroutines.RemoveOnCancel.invoke(CancellableContinuation.kt:297)
	at kotlinx.coroutines.CancellableContinuationImpl.cancel(CancellableContinuationImpl.kt:171)
	... 22 more
```

https://teamcity.jetbrains.com/viewLog.html?buildId=2565287&tab=buildResultsDiv&buildTypeId=KotlinTools_KotlinxCoroutines_BuildWindows
`ChannelAtomicCancelStressTest.testAtomicCancelStress[RendezvousChannel]`
```
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[ThreadPoolDispatcher[2, ChannelAtomicCancelStressTest], Continuation at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchSender$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:108)@1b1b1a34]){CancelledContinuation[kotlinx.coroutines.JobCancellationException: Job was cancelled; job="coroutine#563669":StandaloneCoroutine{Cancelled}@1e724a8a]}@475be4b. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:86)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.DispatchedContinuation
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:106)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:192)
	... 7 more
```

https://teamcity.jetbrains.com/viewLog.html?buildId=2566584&tab=buildResultsDiv&buildTypeId=KotlinTools_KotlinxCoroutines_Build#testNameId-3299188611556527194
`ChannelAtomicCancelStressTest.testAtomicCancelStress[RendezvousChannel]`
```
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[ThreadPoolDispatcher[2, ChannelAtomicCancelStressTest], Continuation at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchSender$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:108)@7920dc2b]){CancelledContinuation[kotlinx.coroutines.JobCancellationException: Job was cancelled; job="coroutine#1525850":StandaloneCoroutine{Cancelled}@3c225d6b]}@a7657db. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:86)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.DispatchedContinuation
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:106)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:192)
	... 7 more
```

https://teamcity.jetbrains.com/viewLog.html?buildId=2567101&tab=buildResultsDiv&buildTypeId=KotlinTools_KotlinxCoroutines_NightlyStress#testNameId-3299188611556527194
`ChannelAtomicCancelStressTest.testAtomicCancelStress[RendezvousChannel]`
```
otlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[ThreadPoolDispatcher[2, ChannelAtomicCancelStressTest], Continuation at kotlinx.coroutines.channels.ChannelAtomicCancelStressTest$launchReceiver$1.invokeSuspend(ChannelAtomicCancelStressTest.kt:135)@73114b46]){CancelledContinuation[kotlinx.coroutines.JobCancellationException: Job was cancelled; job="coroutine#21157923":StandaloneCoroutine{Cancelled}@64e06ece]}@52a44403. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:86)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:57)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to kotlinx.coroutines.DispatchedContinuation
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:106)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:192)
	... 7 more
```

--- a/kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt
+++ b/kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt
@@ -466,7 +466,7 @@ internal abstract class AbstractSendChannel<E> : SendChannel<E> {
                 select.resumeSelectCancellableWithException(closed.sendException)
         }
 
-        override fun toString(): String = "SendSelect($pollResult)[$channel, $select]"
+        override fun toString(): String = "SendSelect@$hexAddress($pollResult)[$channel, $select]"
     }
 
     internal class SendBuffered<out E>(
@@ -476,6 +476,7 @@ internal abstract class AbstractSendChannel<E> : SendChannel<E> {
         override fun tryResumeSend(otherOp: PrepareOp?): Symbol? = RESUME_TOKEN.also { otherOp?.finishPrepare() }
         override fun completeResumeSend() {}
         override fun resumeSendClosed(closed: Closed<*>) {}
+        override fun toString(): String = "SendBuffered@$hexAddress($element)"
     }
 }
 
@@ -899,9 +900,10 @@ internal abstract class AbstractChannel<E> : AbstractSendChannel<E>(), Channel<E
 
         @Suppress("IMPLICIT_CAST_TO_ANY")
         override fun tryResumeReceive(value: E, otherOp: PrepareOp?): Symbol? {
-            otherOp?.finishPrepare()
             val token = cont.tryResume(resumeValue(value), otherOp?.desc) ?: return null
             assert { token === RESUME_TOKEN } // the only other possible result
+            // We can call finishPrepare only after successful tryResume, so that only good affected node is saved
+            otherOp?.finishPrepare()
             return RESUME_TOKEN
         }
 
@@ -914,7 +916,7 @@ internal abstract class AbstractChannel<E> : AbstractSendChannel<E>(), Channel<E
                 else -> cont.resumeWithException(closed.receiveException)
             }
         }
-        override fun toString(): String = "ReceiveElement[receiveMode=$receiveMode]"
+        override fun toString(): String = "ReceiveElement@$hexAddress[receiveMode=$receiveMode]"
     }
 
     private class ReceiveHasNext<E>(
@@ -922,9 +924,10 @@ internal abstract class AbstractChannel<E> : AbstractSendChannel<E>(), Channel<E
         @JvmField val cont: CancellableContinuation<Boolean>
     ) : Receive<E>() {
         override fun tryResumeReceive(value: E, otherOp: PrepareOp?): Symbol? {
-            otherOp?.finishPrepare()
             val token = cont.tryResume(true, otherOp?.desc) ?: return null
             assert { token === RESUME_TOKEN } // the only other possible result
+            // We can call finishPrepare only after successful tryResume, so that only good affected node is saved
+            otherOp?.finishPrepare()
             return RESUME_TOKEN
         }
 
@@ -948,7 +951,7 @@ internal abstract class AbstractChannel<E> : AbstractSendChannel<E>(), Channel<E
                 cont.completeResume(token)
             }
         }
-        override fun toString(): String = "ReceiveHasNext"
+        override fun toString(): String = "ReceiveHasNext@$hexAddress"
     }
 
     private class ReceiveSelect<R, E>(
@@ -983,7 +986,7 @@ internal abstract class AbstractChannel<E> : AbstractSendChannel<E>(), Channel<E
                 channel.onReceiveDequeued() // notify cancellation of receive
         }
 
-        override fun toString(): String = "ReceiveSelect[$select,receiveMode=$receiveMode]"
+        override fun toString(): String = "ReceiveSelect@$hexAddress[$select,receiveMode=$receiveMode]"
     }
 }
 
@@ -1022,7 +1025,7 @@ internal abstract class Send : LockFreeLinkedListNode() {
     // Returns: null - failure,
     //          RETRY_ATOMIC for retry (only when otherOp != null),
     //          RESUME_TOKEN on success (call completeResumeSend)
-    // Must call otherOp?.finishPrepare() before deciding on result other than RETRY_ATOMIC
+    // Must call otherOp?.finishPrepare() after deciding on result other than RETRY_ATOMIC
     abstract fun tryResumeSend(otherOp: PrepareOp?): Symbol?
     abstract fun completeResumeSend()
     abstract fun resumeSendClosed(closed: Closed<*>)
@@ -1036,7 +1039,7 @@ internal interface ReceiveOrClosed<in E> {
     // Returns: null - failure,
     //          RETRY_ATOMIC for retry (only when otherOp != null),
     //          RESUME_TOKEN on success (call completeResumeReceive)
-    // Must call otherOp?.finishPrepare() before deciding on result other than RETRY_ATOMIC
+    // Must call otherOp?.finishPrepare() after deciding on result other than RETRY_ATOMIC
     fun tryResumeReceive(value: E, otherOp: PrepareOp?): Symbol?
     fun completeResumeReceive(value: E)
 }
@@ -1050,14 +1053,15 @@ internal class SendElement(
     @JvmField val cont: CancellableContinuation<Unit>
 ) : Send() {
     override fun tryResumeSend(otherOp: PrepareOp?): Symbol? {
-        otherOp?.finishPrepare()
         val token = cont.tryResume(Unit, otherOp?.desc) ?: return null
         assert { token === RESUME_TOKEN } // the only other possible result
+        // We can call finishPrepare only after successful tryResume, so that only good affected node is saved
+        otherOp?.finishPrepare() // finish preparations
         return RESUME_TOKEN
     }
     override fun completeResumeSend() = cont.completeResume(RESUME_TOKEN)
     override fun resumeSendClosed(closed: Closed<*>) = cont.resumeWithException(closed.sendException)
-    override fun toString(): String = "SendElement($pollResult)"
+    override fun toString(): String = "SendElement@$hexAddress($pollResult)"
 }
 
 /**
@@ -1076,7 +1080,7 @@ internal class Closed<in E>(
     override fun tryResumeReceive(value: E, otherOp: PrepareOp?): Symbol? = RESUME_TOKEN.also { otherOp?.finishPrepare() }
     override fun completeResumeReceive(value: E) {}
     override fun resumeSendClosed(closed: Closed<*>) = assert { false } // "Should be never invoked"
-    override fun toString(): String = "Closed[$closeCause]"
+    override fun toString(): String = "Closed@$hexAddress[$closeCause]"
 }
 
 private abstract class Receive<in E> : LockFreeLinkedListNode(), ReceiveOrClosed<E> {

--- a/kotlinx-coroutines-core/jvm/src/internal/LockFreeLinkedList.kt
+++ b/kotlinx-coroutines-core/jvm/src/internal/LockFreeLinkedList.kt
@@ -399,7 +399,7 @@ public actual open class LockFreeLinkedListNode {
 
         // Returns REMOVE_PREPARED or null (it makes decision on any failure)
         override fun perform(affected: Any?): Any? {
-            assert(affected === this.affected)
+            assert { affected === this.affected }
             affected as Node // type assertion
             val decision = desc.onPrepare(this)
             if (decision === REMOVE_PREPARED) {

